### ISSUE TITLE
nunit3-console.exe hangs when running in parallel by multiple processes

### ISSUE THREAD

**Issue by justsomeuselessguy** [nunit/nunit-console#663]
nunit3-console.exe hangs from time to time when running in parallel by multiple processes. I was trying to debug it with dnSpy and it looks like all threads are waiting for some event which doesn't set. Memory dump zip file: https://drive.google.com/file/d/1f-bUBWgq8ljZrp04rEaHwbzLrg4w-_zW/view?usp=sharing

**Comment by ChrisMaddock** [comment]
There's been a fix to a similar issue in master recently, would you try running on the latest build of master, and see if you still see the hangs? You can find the build on MyGet

https://www.myget.org/feed/nunit/package/nuget/NUnit.ConsoleRunner

**Comment by justsomeuselessguy** [comment]
> There's been a fix to a similar issue in master recently, would you try running on the latest build of master, and see if you still see the hangs? You can find the build on MyGet
> 
> https://www.myget.org/feed/nunit/package/nuget/NUnit.ConsoleRunner

I still can reproduce this issue using NUnit.ConsoleRunner.3.11.0-dev-04532.

**Comment by ChrisMaddock** [comment]
Ok. Could you run with —trace=debug please, and share the logs from when it hangs?

**Comment by ChrisMaddock** [comment]
(On that myget build of master is probably best - rule out any already solved problems!)

**Comment by justsomeuselessguy** [comment]
[InternalTrace.13132.log](https://github.com/nunit/nunit-console/files/3382355/InternalTrace.13132.log)
[InternalTrace.13132.nunit.framework.tests2.dll.log](https://github.com/nunit/nunit-console/files/3382356/InternalTrace.13132.nunit.framework.tests2.dll.log)

**Comment by CharliePoole** [comment]
You appear to be running a modified set of NUnit's own tests. Is that correct? What mods have you made?

**Comment by justsomeuselessguy** [comment]
I have just make a copy of original nunit.framework.tests.dll and rename it.

**Comment by CharliePoole** [comment]
Since all those tests run OK in our CI, I'm wondering if you are using some different options to run them. What is your command-line?

**Comment by justsomeuselessguy** [comment]
X:\instal\NUnitConsole\packages\NUnit.ConsoleRunner.3.11.0-dev-04532\tools\nunit3-console.exe --result:"a3abbda9-d1dc-4546-9212-1cb3999f44a5.xml;format=nunit3" "X:/instal\\NUnit.Framework-3.11.0\\bin\\net40\\nunit.framework.tests.dll" & ... and so on in a cycle.

**Comment by CharliePoole** [comment]
Also, where do the multiple processes come in? I see one test process in the log with multiple threads.

**Comment by justsomeuselessguy** [comment]
I have run this command in multiple PowerShell windows

**Comment by justsomeuselessguy** [comment]
Some are hangs, other aren't. This is just a simple way of reproducing this issue without solution that I'm working on. It is for parallel execution of unit tests by remote machines across the network.

**Comment by CharliePoole** [comment]
OK... I was suspecting something like that. For that to be valid, you would need to run a test assembly that was designed to be run multiple times in parallel. Our tests are not. For example, files are accessed on occasion. In order to run successfully, we make such tests non-parallelizable, but that only refers to parallelism within a single process and assembly. If you are running other processes at the same time, you are isolated WRT in-memory stuff but not external things.

I'm also seeing some problem with the tests that use compilation (ShouldNotCompile) it's quite likely that the old technology those tests are using can't be run more than once at a time.

Basically, our own framework tests are exactly good enough to do what they need to do and no more. :smile: What you're doing with them is beyond their intent. In your place, I'd create (or copy) specific tests that are clearly isolated so that they can run in multiple copies in different processes.

**Comment by CharliePoole** [comment]
I get it's just an easy way to test, but I think it may introduce other stuff that obfuscates the main issue.

**Comment by justsomeuselessguy** [comment]
Well, thank you then for a clarify answer.

**Comment by seangriffin2** [comment]
> There's been a fix to a similar issue in master recently, would you try running on the latest build of master, and see if you still see the hangs? You can find the build on MyGet
> 
> https://www.myget.org/feed/nunit/package/nuget/NUnit.ConsoleRunner

Hey @ChrisMaddock - Is this fix related to a specific Issue Number? I am just curious because I am not sure if I am running into the same problem on **NUnit Console Runner 3.10.0 (.NET 2.0)** . Thank you guys for all the hard work!

Sean

**Comment by ChrisMaddock** [comment]
Hi @seangriffin2 - try 3.11. If that doesn't help, I'd suggest opening a new issue, as I expect you have a different problem to the one discussed here. (Unless you're also running NUnit's own tests! 🙂)

**Comment by ChrisMaddock** [comment]
Closing this, as I think the questions answered. Feel free to comment further if not, the team will still get the notifications. 🙂
